doc/rst/platforms/comm-layers/gasnet.rst

```diff
@@ -144,6 +162,26 @@ on a single machine with the following environment variable settings:
 See the documentation for the :ref:`UDP conduit <using-udp>` for more details on
 what these environment variables do.
 
+.. _readme-udp-vs-smp:
+
+When should I use ``CHPL_COMM_SUBSTRATE=udp`` vs ``CHPL_COMM_SUBSTRATE=smp``?
+*****************************************************************************
+
+While both the ``udp`` and ``smp`` conduits can be used to run multilocale
+Chapel programs on a single node or shared memory machine (e.g. laptops or
+workstations), there are a few key differences:
+
+* ``udp`` emulates multiple locales by :ref:`oversubscribing
+  <oversubscribed-execution>` the machine's resources, so each locale is
+  sharing the same resources. ``smp`` creates multiple co-locales which
+  partition the machine's resources between them, so each locale has its own
+  dedicated resources.
```

Comment:
As a side note, I'd really like udp-conduit to eventually include _real_ support for co-locales (i.e. stop conflating LPN > 1 with oversubscription).

For the single-node case, I believe smp-conduit usually remains the best choice for most use cases. However for a multi-node Ethernet-connected cluster, a user might reasonably want to run one co-locale per socket (or per other arch feature) while using udp-conduit, for all the same reasons it makes sense on other networks.

So while I acknowledge the paragraph above probably represents the _current_ implementation behavior, I sincerely hope this is not permanent. This issue is probably orthogonal to this PR.

Reply:
I think that's a separate issue. Once the udp conduit supports co-locales, we won't stop supporting oversubscription right?

* `-nl 2`: 2 locales, sharing the resources of the machine
* `-nl 1x2`: 2 colocales, partitioning the resources of the machine
* `-nl 2x2`: 4 locales, each using half the machines resources (2 locales per partition)

Reply:
* `-nl 2`: 2 locales, sharing the resources of the machine

The first one should only be oversubscribed if the spawner happens to place both processes on the same physical node (controlled by other knobs elsewhere). Single-node execution of udp-conduit is a degenerate case, which as this list explains is usually better served by smp-conduit. In the general case udp-conduit is multi-node, and that's its expected primary use case.

In any case, we agree this topic is orthogonal to this PR and probably merits further discussion elsewhere.